Enable support for Pointer Authentication for TAs

ruchi393 opened the pull request:
Pointer Authentication (part of ARMv8.3 Extensions) provides instructions for signing and authenticating pointers against secret keys. These can be used to mitigate ROP (Return oriented programming) attacks. The patches enables these instructions for TA's at EL0. When this option is chosen, TEE core will initialise secret keys per TA at the load time. 

The PAUTH support in ARMv8.3 introduces 2 keys for instruction space (APIA/B), 2 for data space (APDA/B) and 1 for general use (APG). The patchset uses only one key i.e APIAKey for TA's.

All xtests along with GP tests pass with the patch-set.

Changes have been done in thread_a64.S to save and restore the keys. Currently it is assumed we are enabling pauth when transitioning to el0 and disabling it whenever there is a sys call or exception return to el1.

To test this, CTX_INCLUDE_PAUTH_REGS needs to be enabled when compiling TF-A to allow for saving/restoring of PAUTH key registers at context switch in TF-A and enabling access of these registers by EL1.

etienne-lms reviewed (commented):
misc minor comments

ruchi393 commented on core/arch/arm/arm.mk:
```diff
@@ -299,12 +299,23 @@ ta_arm64-platform-cxxflags += -fpic
 ta_arm64-platform-cxxflags += $(platform-cflags-optimization)
 ta_arm64-platform-cxxflags += $(platform-cflags-debug-info)
 
+ifeq ($(CFG_TA_PAUTH_ENABLE),y)
```
This is y .
If only PAUTH is enabled -> -mbranch-protection=pac-ret+leaf
If only BTI is enabled -> -mbranch-protection=bti
If both PAUTH and BTI are enabled -> -mbranch-protection=pac-ret+leaf+bti

ruchi393 replied:
Oh Ok..got it..will remove the _ENABLE

ruchi393 commented on core/arch/arm/kernel/link.mk:
```diff
@@ -19,7 +19,7 @@ ifeq ($(CFG_CORE_BTI),y)
 # with the --fatal-warnings below to check and prevent this situation (with useful
 # diagnostics).
-link-ldflags += $(call ld-option,-z force-bti) --fatal-warnings
+#link-ldflags += $(call ld-option,-z force-bti) --fatal-warnings
```
Sorry, need to remove this. We need to have a BTI enabled gcc toolchain. I was using this to test it locally at my end

ruchi393 commented on core/arch/arm/include/arm64.h:
```diff
@@ -199,6 +209,34 @@
+#define ID_AA64ISAR1_API_IMP_DEF_EPAC2_FPAC_CMB	U(0x5)
+
+#define ID_AA64ISAR1_APA_SHIFT			U(4)
+#define ID_AA64ISAR1_API_MASK			U(0xf)
```
My bad. I will change it

jenswi-linaro reviewed (commented):
Comments for "arm64: Add registers and bits for pointer authentication support"

jenswi-linaro commented on core/kernel/user_ta.c:
```diff
@@ -163,9 +165,15 @@ static TEE_Result user_ta_enter(struct ts_session *session,
 	res = thread_enter_user_mode(func, kaddr_to_uref(session),
 				     (vaddr_t)usr_params, cmd, usr_stack,
 				     utc->uctx.entry_func, utc->uctx.is_32bit,
+				     key_hi, key_lo,
```
We're accumulating quite a bit of arguments for `thread_enter_user_mode()`. How about replacing everything after the `entry_func` parameter with `utc` and let `thread_enter_user_mode()` dig out the rest of the parameters there? In the process we'll get rid of a bunch of `#ifdef CFG_TA_PAUTH_ENABLE`.

etienne-lms replied:
+1

ruchi393 replied:
thread_enter_user_mode() is also called from core/arch/arm/kernel/ldelf_loader.c where utc is not available. I can try passing uctx (i.e  struct user_mode_ctx). It has is_32bit in it and also pointer to ts_ctx from where keys can be derived. But would this work with core/arch/arm/kernel/ldelf_loader.c where when we are calling thread_enter_user_mode(), we use is_arm32 which is globally defined. I think I might be missing something here.

jenswi-linaro replied:
Hmm, that was unfortunate. How about passing the two keys as an array where the pointer is retrieved with a helper function that just returns NULL if CFG_TA_PAUTH_ENABLE isn't enabled? That should at least save us one argument and a bunch of ifdefs.

ruchi393 replied:
I am making some changes to remove the key parameter from this function. Instead we will get teh current session and get the keys from the ts_ctx. Please have a look

jenswi-linaro commented on core/include/kernel/ts_manager.h:
```diff
@@ -13,6 +13,10 @@
 struct ts_ctx {
 	TEE_UUID uuid;
+#ifdef CFG_TA_PAUTH_ENABLE
+	uint64_t pauth_key_hi;
```
This should perhaps be `uint64_t pauth_key[2];` instead to allow a single call to `crypto_rng_read()` in `tee_ta_init_user_ta_session()`

jenswi-linaro reviewed (commented):
Some comments for "Add basic pointer authentication support for TA's"

ruchi393 commented on core/arch/arm/include/arm64.h:
```diff
@@ -135,11 +141,15 @@
 #define ESR_EC_AARCH32_CP14_LS	U(0x06)
 #define ESR_EC_FP_ASIMD		U(0x07)
 #define ESR_EC_AARCH32_CP10_ID	U(0x08)
+#define ESR_EC_PAC		U(0x09)
```
Changed

jenswi-linaro reviewed (commented):
Do we need to clear the apiakey before returning back to normal world, or is that a TF-A problem?

ruchi393 commented:
> Do we need to clear the apiakey before returning back to normal world, or is that a TF-A problem?

This is taken care by TF-A. It is responsible for saving and restoring these keys during context switch.

ruchi393 commented on core/include/kernel/ts_manager.h:
```diff
@@ -13,6 +13,9 @@
 
 struct ts_ctx {
 	TEE_UUID uuid;
+#ifdef CFG_TA_PAUTH
```
Ok. I will shift it there.

etienne-lms replied:
don't forget to remove this change

ruchi393 commented on core/include/kernel/tee_ta_manager.h:
```diff
@@ -157,6 +157,12 @@ static inline void tee_ta_ftrace_update_times_suspend(void) {}
 #endif
 
+#if defined(CFG_TA_PAUTH)
+void tee_ta_get_pauth_key(unsigned long *key);
```
Ok

ruchi393 commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1479,11 +1493,11 @@ uint32_t thread_enter_user_mode(unsigned long a0, unsigned long a1,
 	tee_ta_update_session_utime_resume();
 
-	tee_ta_get_pauth_key(keys);
+	keys = thread_get_pauth_keys();
```
To get the pauth keys in the thread_enter_user_mode function, I have 2 choices ->
one which I have implemented here i.e reach user_mode_ctx from ts_get_current_session()
or 
-> uctx is known in all the places from where thread_enter_user_mode() is called from i,e user_ta.c and ldelf_loader.c, So add an argument to thread_enter_user_mode which is uctx

Please suggest what would be better thing to do.

jenswi-linaro replied:
The approach here looks good to me.

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1461,11 +1461,29 @@ static void set_ctx_regs(struct thread_ctx_regs *regs, unsigned long a0,
+	/* Only user TA's support the PAUTH keys */
+	struct user_ta_ctx *utc = to_user_ta_ctx(s->ctx);
+
+	return &(utc->uctx.keys);
```
Please drop the `()`, they are not needed.

ruchi393 commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -640,6 +682,14 @@ LOCAL_FUNC el0_svc , :
 	store_xregs sp, THREAD_SVC_REG_ELR, 0, 1
+
+#ifdef CFG_TA_PAUTH
+	/* Save APIAKEY */
```
Without saving and restoring these registers at svc time,  I was getting a lot of xtest errors where the authentication was failing.  IIRC, most of these tests were the ones which were using storage and had some RPC calls. I will re-check and get back on this.

etienne-lms commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1461,11 +1461,29 @@ static void set_ctx_regs(struct thread_ctx_regs *regs, unsigned long a0,
 
+static struct thread_pauth_keys *thread_get_pauth_keys(void)
+{
+#ifdef CFG_TA_PAUTH
```
if think using `if (IS_ENABLED(CFG_TA_PAUTH)) return NULL;` should build ok.

ruchi393 replied:
This doesn't build fine as struct user_mode_ctx’ has no member named ‘keys’ if CFG_TA_PAUTh is not enabled.

etienne-lms commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -677,6 +727,8 @@ LOCAL_FUNC el0_svc , :
 	msr	elr_el1, x0
 	msr	spsr_el1, x1
+
+
```
maybe discard this change

etienne-lms commented on lib/libunw/include/unw/unwind.h:
```diff
@@ -118,4 +118,6 @@ static inline void print_stack_arm64(struct unwind_state_arm64 *state __unused,
  */
 void ftrace_map_lr(uint64_t *lr);
 
+void pauth_strip_pac(uint64_t *lr);
```
a word of description would be nice.
/* Strip out PAuth tags from LR content if applicable */

etienne-lms commented on mk/config.mk:
```diff
@@ -776,3 +776,12 @@ CFG_CORE_ASYNC_NOTIF ?= n
+# instructions for signing and authenticating pointers against secret keys.
+# These can be used to mitigate ROP (Return oriented programming) attacks.
+# This option enables these instructions for TA's at EL0. When this option is
+# chosen, TEE core will initialise secret keys per TA.
```
s/chosen/enabled/
s/initialise/initialize/

etienne-lms reviewed (commented):
some minor comments

ruchi393 commented:
Updated PR. Fixed comments till date.

2 new commits have been added
- A property to check PAuth in TEE property set
-  CFG_TA_PAUThH is exported for TA's.

There are many review commits in the PR now. For easier review, if you are ok I can squash them together,

jenswi-linaro commented:
> There are many review commits in the PR now. For easier review, if you are ok I can squash them together,

Please do so.

ruchi393 commented:
PR updated with patches squashed for better review.

etienne-lms commented:
LGTM. `Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for the 6 commits.

ruchi393 commented:
Thanks @etienne-lms .

@jenswi-linaro ,  can you have a look at the updated PR.

jenswi-linaro commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -52,7 +74,15 @@ FUNC thread_resume , :
+#ifdef	CFG_TA_PAUTH
+	/* Restore PAC keys before return to el0 */
+	load_xregs x0, THREAD_CTX_REGS_APIAKEY_HI, 1, 2
+	msr	S3_0_c2_c1_1, x1
```
Please add read and write macros for these registers. See <out>/core/include/generated/arm32_sysreg.S for examples. In this case I guess we can hand code them and add them to core/arch/arm/include/arm64_macros.S.

jenswi-linaro commented on ldelf/pauth.c:
```diff
@@ -0,0 +1,17 @@
+#include <util.h>
+#include "pauth.h"
+
+#ifdef CFG_TA_PAUTH
```
This isn't needed as this file is only compiled if `CFG_TA_PAUTH=y`

jenswi-linaro commented on ldelf/pauth.h:
```diff
@@ -0,0 +1,17 @@
+#include <types_ext.h>
+#include <util.h>
+
+#ifdef CFG_TA_PAUTH
```
This is overkill, we'll get a link error anyway if this function is used.

jenswi-linaro commented on core/arch/arm/include/arm.h:
```diff
@@ -125,6 +125,22 @@ static inline bool feat_bti_is_implemented(void)
+#ifdef ARM32
+	return false;
+#else
+	uint64_t mask = (ID_AA64ISAR1_GPI_MASK << ID_AA64ISAR1_GPI_SHIFT) |
```
`SHIFT_U64()` is more safe to use.

ruchi393 commented:
PR updated based on comments.

etienne-lms commented on core/arch/arm/include/arm.h:
```diff
@@ -131,10 +131,11 @@ static inline bool feat_pauth_is_implemented(void)
-			(ID_AA64ISAR1_GPA_MASK << ID_AA64ISAR1_GPA_SHIFT) |
-			(ID_AA64ISAR1_API_MASK << ID_AA64ISAR1_API_SHIFT) |
-			(ID_AA64ISAR1_APA_MASK << ID_AA64ISAR1_APA_SHIFT);
+	uint64_t mask =
```
could add `const`

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for the 3 fixup commits.

jenswi-linaro reviewed (approved):
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

ruchi393 commented:
Thanks @jenswi-linaro and @etienne-lms for the review.

@etienne-lms , I will be squashing the review commits with the other patches in the PR. Earlier you had given a A-by tag for the commits and you mentioned a R-by tag for the review commits which will be squashed. Can you please confirm if you would like me to use A-by tag or the R-by tag ?

etienne-lms commented:
Sorry for the confusion. My  `Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for the series.

ruchi393 commented:
PR updated with tags.

ruchi393 commented:
PR rebased to fix merge conflict.

@jforissier , can you pick this up.

jforissier commented:
@ruchi393 thanks for the update. I will wait for CI to finish as a matter of course then I will merge this.

jforissier commented:
With this PR merged, the QEMUv8 build doesn't boot properly when `CFG_TA_PAUTH=y`. It looks like secure world hangs trying to run ldelf. Did I miss something?
```
$ make -j10 CFG_TA_PAUTH=y run

NS log:

[    4.479864] optee: probing for conduit method.
[    4.482365] optee: revision 3.16 (f0d8ad67)
[    4.489616] optee: dynamic shared memory is enabled
[    4.511449] optee: initialized driver
[    4.515191] Driver 'optee' was unable to register with bus_type 'arm_ffa' because the bus was not initialized.
[    4.558215] NET: Registered PF_PACKET protocol family
[    4.563313] 9pnet: Installing 9P2000 support
[    4.565114] Key type dns_resolver registered
[    4.570576] Loading compiled-in X.509 certificates
[    4.583732] pstore: Using crash dump compression: deflate
[   25.598071] rcu: INFO: rcu_preempt detected stalls on CPUs/tasks:
[   25.599710] rcu:     1-...0: (1 GPs behind) idle=66f/1/0x4000000000000000 softirq=278/279 fqs=2625
[   25.601404]  (detected by 0, t=5252 jiffies, g=-571, q=4)
[   25.602696] Task dump for CPU 1:
[   25.603319] task:swapper/0       state:R  running task     stack:    0 pid:    1 ppid:     0 flags:0x0000000a
[   25.606116] Call trace:
[   25.606642]  __switch_to+0x13c/0x194
[   25.608376]  0x92009e7ee1bd6b00

S log:

I/TC: Primary CPU initializing
D/TC:0 0 boot_init_primary_late:1302 Executing at offset 0x4c031000 with virtual load address 0x5a131000
D/TC:0 0 call_preinitcalls:21 level 2 mobj_mapped_shm_init()
D/TC:0 0 mobj_mapped_shm_init:463 Shared memory address range: 5b200000, 5d200000
D/TC:0 0 call_initcalls:40 level 1 register_time_source()
D/TC:0 0 call_initcalls:40 level 1 teecore_init_pub_ram()
D/TC:0 0 call_initcalls:40 level 2 probe_dt_drivers_early()
D/TC:0 0 call_initcalls:40 level 3 check_ta_store()
D/TC:0 0 check_ta_store:408 TA store: "early TA"
D/TC:0 0 check_ta_store:408 TA store: "Secure Storage TA"
D/TC:0 0 check_ta_store:408 TA store: "REE"
D/TC:0 0 call_initcalls:40 level 3 early_ta_init()
D/TC:0 0 early_ta_init:56 Early TA f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c size 33252 (compressed, uncompressed 55144)
D/TC:0 0 call_initcalls:40 level 3 verify_pseudo_tas_conformance()
D/TC:0 0 call_initcalls:40 level 3 tee_cryp_init()
D/TC:0 0 call_initcalls:40 level 4 tee_fs_init_key_manager()
D/TC:0 0 call_initcalls:40 level 5 probe_dt_drivers()
D/TC:0 0 call_initcalls:40 level 6 mobj_init()
D/TC:0 0 call_initcalls:40 level 6 default_mobj_init()
D/TC:0 0 call_initcalls:40 level 7 release_probe_lists()
D/TC:0 0 call_finalcalls:59 level 1 release_external_dt()
I/TC: Primary CPU switching to normal world boot
I/TC: Secondary CPU 1 initializing
I/TC: Secondary CPU 1 switching to normal world boot
D/TC:0   tee_entry_exchange_capabilities:100 Asynchronous notifications are disabled
D/TC:0   tee_entry_exchange_capabilities:109 Dynamic shared memory is enabled
D/TC:0 0 core_mmu_xlat_table_alloc:513 xlat tables used 7 / 8
D/TC:? 0 tee_ta_init_pseudo_ta_session:296 Lookup pseudo TA 7011a688-ddde-4053-a5a9-7b3c4ddf13b8
D/TC:? 0 tee_ta_init_pseudo_ta_session:309 Open device.pta
D/TC:? 0 tee_ta_init_pseudo_ta_session:326 device.pta : 7011a688-ddde-4053-a5a9-7b3c4ddf13b8
D/TC:? 0 tee_ta_close_session:512 csess 0x5a1ba9a0 id 1
D/TC:? 0 tee_ta_close_session:531 Destroy session
D/TC:? 0 tee_ta_init_pseudo_ta_session:296 Lookup pseudo TA f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c
D/TC:? 0 ldelf_load_ldelf:95 ldelf load address 0x80006000
```

ruchi393 commented:
@jforissier, for this to work :
- CTX_INCLUDE_PAUTH_REGS=1  needs to be used when compiling TF-A to allow for saving/restoring of PAUTH key registers at context switch in TF-A and enabling access of these registers by EL1.

jforissier commented:
@ruchi393 oh, yes, you mentioned this in the PR description :-/ sorry about that. It worked great after I added the flags. IMO build.git needs to be updated.

So when TF-A isn't configured properly, I guess some pointer authentication fails and there is some kind of exception, right? Could we have a crash dump in this case? (like when the TA performs a data-abort etc.).

ruchi393 commented:
> @ruchi393 oh, yes, you mentioned this in the PR description :-/ sorry about that. It worked great after I added the flags. IMO build.git needs to be updated.
I do plan to send out a patch to build.git for this.
> 
> So when TF-A isn't configured properly, I guess some pointer authentication fails and there is some kind of exception, right? Could we have a crash dump in this case? (like when the TA performs a data-abort etc.).
Here the failure is not due to pointer authentication. Access to the registers holding "key" values for pointer authentication from EL1 or EL2  or use of any instruction related to pointer authentication in any Exception level except EL3 are trapped to EL3 if the API, APK bit is not set in SCR_EL3. These bits are set by TF-A only when CTX_INCLUDE_PAUTH_REGS=1.

jforissier commented:
@ruchi393 got it, thanks.
